Add hooks for dash and related packages (#102)

KyleKing opened the pull request:
Fixes #102

So I started with `dash_html_components` which was my initial use case, but since getting positive feedback, I did some additional research to see what users were struggling with ([here](https://stackoverflow.com/questions/59914106/pyinstaller-error-when-executing-plotly-dash-exec-file), [here](https://stackoverflow.com/questions/60049673/how-to-pack-a-plotly-dash-application-with-pyinstaller-into-a-single-exe-file-fo), etc.). I also did a deeper dive using a standalone project to test out the hooks and make sure they covered the most common use cases: https://github.com/KyleKing/Poetry-EdgeCaseTesting

There is also an issue with `dash` which calls `pkg_resources.get_distribution` on `flask-compress` so there needs be a `copy_metadata` in the flask compress hook

I would be happy to incorporate any feedback. I think I manually tested all of the changes, but I could have overlooked something since PyInstaller is fairly complex

> Side note: What happens if someone needs to override one of these hooks? Does the hook from this package get called first and the user overrides it? I see a warning for `win32ctypes.core` (`311 WARNING: Several hooks defined for module 'win32ctypes.core'. Please take care they do not conflict.`)
>
> *Update*: [From the docs](https://pyinstaller.readthedocs.io/en/stable/hooks.html#providing-pyinstaller-hooks-with-your-package) "If both PyInstaller and your package provide hooks for some module, your package’s hooks take precedence, but can still be overridden by the command line option --additional-hooks-dir." (i.e. pyinstaller > package > user)

rokm commented on src/_pyinstaller_hooks_contrib/hooks/stdhooks/hook-flask_compress.py:
```diff
@@ -0,0 +1,15 @@
+
+from PyInstaller.utils.hooks import copy_metadata
+
+datas = (copy_metadata('flask_compress'))
```
Those parentheses around copy_metadata should be redundant.

rokm commented on src/_pyinstaller_hooks_contrib/hooks/stdhooks/hook-plotly.py:
```diff
@@ -0,0 +1,16 @@
+from PyInstaller.utils.hooks import collect_data_files
+
+datas = collect_data_files('plotly', include_py_files=True,
+                           includes=['**/*.py', 'package_data/**/*.*'])
```
Do we really need `include_py_files=True` and explicit collection of `.py` files?

I see from plotly/plotly.py#3122 that this is probably an attempt to work-around the lazy imports, but the proper way to deal with those is to collect problematic submodules into `hiddenimports`, i.e.:
```python
from PyInstaller.utils.hooks import collect_submodules
hiddenimports = collect_submodules('plotly.validators')
```

KyleKing replied:
Okay I didn’t know PyInstaller could use that. I’ll test it out this evening and resubmit if it works!

rokm commented:
Also, please add a [news fragment](https://github.com/pyinstaller/pyinstaller-hooks-contrib/blob/master/news/README.txt) for the PR.

bwoodsend commented:
It would be good if you can provide a test for each one too. Just the least amount of code required to cause PyInstaller to break without your new hooks. Simple hello world examples from their docs are ideal.

KyleKing commented:
I think everything should be ready to go. I had to wrap up some school work, but had some time this weekend to revisit

- [x] Implemented recommendations to removed excess parenthesis and use `collect_submodules()`
- [x] Added new fragment `103.new.rst`
- [x] May not be within scope for this PR, but I incremented h5py and av for Python >3.9 to latest versions that have pre-built Python 3.9 wheels to fix test failures
- [x] Added a test for `dash`, which also tests `dash-core-components`, `dash-html-components`, `dash_renderer`, and `flask-compress`
- [x] Added test for `dash_table`
- [x] Added test for `dash_bootstrap_components`
- [x] *Removed* `dash_extensions`. This is package is a bit more complex and I would want to create a separate PR
- [x] Added test for `plotly` that would fail if not for the hooks
    - I didn't include the CSV data in plotly, which means that `px.iris()` would fail, but including them should probably be up to the user so they can limit the executable size

bwoodsend commented:
OK getting there. Commit history is a bit of a mess though. Do you know how to edit git history with `git rebase`? Ideally it should be 3 commits:

1. Repinning `h5py` and `av` to appease CI.
2. Add the dask stuff. Hooks, tests, entries in requirements.txt, news fragment (with a fullstop on the end of the fragment's text please!).
3. Add the plotly stuff. Again the hook, the test, the requirement and its own news fragment.

When you do number 1, could you also insert `--prefer-binary` to [this line here](https://github.com/pyinstaller/pyinstaller-hooks-contrib/blob/b3be175808ab5eb11f36ec952aea2c4010c53a02/.github/workflows/macos-test.yml#L37) before the `-r` which should discourage it from trying (and failing) to compile everything from source.

KyleKing commented:
No problem, I can force push cleaned up changes today. It looks like the only remaining issue with the CI was pandas 1.2 which dropped Python 3.6 (I only tested 3.7 and 3.9 locally). I'll include pandas 1.1 when I push the edited history

I didn't see a commit style guide, but I modeled these after other ones in commit history. Any amends?

- CI/CD: increment h5py and av for 3.9 binaries
- Hooks: add hooks and test for plotly
- Hooks: add hooks and tests for dash
- Hooks: add hook and test dash_bootstrap_components

rokm commented on requirements-test-libraries.txt:
```diff
@@ -19,11 +22,15 @@ pyzmq==19.0.0
 # ------------------- Python Version/Platform (OS) specifics
+# No prebuilt wheels for Python 3.9
+h5py==2.10.0; python_version < '3.9'
+h5py==3.2.1; python_version > '3.9'
```
One of these two should probably include '='.

KyleKing replied:
Yeah, caught that locally

bwoodsend commented:
> I didn't see a commit style guide, but I modeled these after other ones in commit history. Any amends?

Honestly I don't really care - as long as it's self explanatory. Those look fine to me.

bwoodsend commented on news/103.new.rst:
```diff
@@ -0,0 +1,3 @@
+Add hook for ``plotly`` to collect data files and hidden `pandas`, `cmath`, and `plotly.validator` imports
+Add hooks for ``dash`` and related packages to collect data files and hook for meta-data from ``flask-compress``
+Add hook for ``dash_bootstrap_components`` to collect data files
```
RestructuredText ignores single line breaks so these sentences will all render together into one long sentence. You'll need to add an extra blank line between each one.
```suggestion
Add hook for ``plotly`` to collect data files and hidden `pandas`, `cmath`, and `plotly.validator` imports

Add hooks for ``dash`` and related packages to collect data files and hook for meta-data from ``flask-compress``

Add hook for ``dash_bootstrap_components`` to collect data files
```

KyleKing commented:
I think the remaining CI issues are unrelated to the Dash changes. All of the macOS CI passed!

On [AppVeyor](https://ci.appveyor.com/project/PyInstaller/pyinstaller-hooks-contrib/builds/38549031), installing H5Py failed for both 3.9 environments. Looks like there wasn't a [Win32 wheel in the 3.2.1 H5Py releases](https://pypi.org/project/h5py/3.2.1/#files), but there is one in [the 3.2.0 release](https://pypi.org/project/h5py/3.2.0/#files). Decrementing down to 3.2.0 might fix this failure. Not sure if the missing wheel is just a release error or not

```sh
Loading library to get build settings and version: hdf5.dll
error: Unable to load dependency HDF5, make sure HDF5 is installed properly
error: Could not find module 'hdf5.dll' (or one of its dependencies). Try using the full path with constructor syntax.
----------------------------------------
ERROR: Failed building wheel for h5py
```

The `passlib` tests on [Travis](https://travis-ci.org/github/pyinstaller/pyinstaller-hooks-contrib/jobs/765870992) failed only on 3.9, but the rest of the development tests were canceled. I think the error was from a known issue in `passlib` (https://foss.heptapod.net/python-libs/passlib/-/issues/115) patched in ~Apr2020 (tested `passlib` version was [1.7.2 from 22Nov2019](https://pypi.org/project/passlib/#history)). This error can probably be fixed by incrementing `passlib` to `1.7.4` for `>= 3.9`

```sh
...
src/_pyinstaller_hooks_contrib/tests/test_libraries.py ssssssssssXXXXXXs [  8%]
s..........ssssss....................ss..ss......FF..sssssssssssssssssss [ 43%]
ssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssssss [ 78%]
sssssssssssssssssssssssssssssssssssss........                            [100%]

=================================== FAILURES ===================================

_____________________________ test_passlib[onedir] _____________________________

pyi_builder = <PyInstaller.utils.conftest.AppBuilder object at 0x7fd6cc12d220>
    @importorskip('passlib')
    def test_passlib(pyi_builder):
>       pyi_builder.test_source("""
            import passlib.apache
            """)
E       Failed: Running exe /tmp/pytest-of-travis/pytest-0/test_passlib_onedir_0/dist/test_source/test_source failed with return-code 255.

/home/travis/build/pyinstaller/pyinstaller-hooks-contrib/src/_pyinstaller_hooks_contrib/tests/test_libraries.py:350: Failed

----------------------------- Captured stdout call -----------------------------

...

[6158] LOADER: Running test_source.py

Traceback (most recent call last):
  File "test_source.py", line 3, in <module>
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
  File "PyInstaller/loader/pyimod03_importers.py", line 531, in exec_module
  File "passlib/apache.py", line 476, in <module>
  File "passlib/apache.py", line 440, in _init_default_schemes
  File "passlib/registry.py", line 538, in has_os_crypt_support
  File "passlib/registry.py", line 491, in has_backend
  File "passlib/utils/handlers.py", line 2111, in has_backend
  File "passlib/utils/handlers.py", line 2189, in set_backend
  File "passlib/utils/handlers.py", line 2312, in _set_backend
  File "passlib/utils/handlers.py", line 2225, in _set_backend
  File "passlib/handlers/bcrypt.py", line 719, in _load_backend_mixin
  File "passlib/utils/__init__.py", line 851, in test_crypt
  File "passlib/utils/__init__.py", line 794, in safe_crypt
  File "crypt.py", line 82, in crypt
OSError: [Errno 22] Invalid argument

...
```

bwoodsend commented:
Yeah you're right. I just forgot to merge it.
